bin/emqx

```diff
@@ -464,28 +467,52 @@ if [ "$IS_BOOT_COMMAND" = 'yes' ]; then
         export EMQX_BOOT_CONFIGS
     fi
 else
-    # For non-boot commands, we try to get data_dir and ssl_dist_optfile from 'ps -ef' output
+    # Turn off debug as the ps output can be quite noisy
+    set +x
+    # For non-boot commands, we need below runtime facts to connect to the running node:
+    #  1. The running node name.
+    #  2. The Erlang cookie in use by the running node name.
+    #  3. SSL options if the node is using TLS for Erlang distribution
+    #
+    # There are 3 sources of truth to get those runtime information.
+    # Listed in the order of preference:
+    #  1. The boot command (which can be inspected from 'ps -ef' command output)
+    #  2. The generated app.<time>.config file located in the dir pointed by 'node.data_dir'
+    #  3. The bootstrap config 'etc/emqx.conf'
+    #
+    # NOTES:
+    #  * We should avoid getting runtime information with the 3rd approach because 'etc/emqx.conf' might
+    #    be updated after the node is started. e.g. If a user starts the node with name 'emqx@127.0.0.1'
+    #    then update the config in the file to 'node.name = "emqx@local.net"', after this change,
+    #    there would be no way stop the running node 'emqx@127.0.0.1', because 'emqx stop' command
+    #    would try to stop the new node instead.
+    #  * The primary grep pattern is $RUNNER_ROOT_DIR because one can start multiple nodes at the same time
+    #  * The grep args like '[e]mqx' but not 'emqx' is to avoid greping the grep command itself
+    #  * The running 'remsh' and 'nodetool' processes must be excluded
     # shellcheck disable=SC2009
-    PS_LINE="$(ps -ef | grep "\-[r]oot $RUNNER_ROOT_DIR" || true)"
+    PS_LINE="$(ps -ef | grep '[e]mqx' | grep -v -E '(remsh|nodetool)' | grep -oE "\-[r]oot ${RUNNER_ROOT_DIR}.*" || true)"
+    [ "$DEBUG" -eq 1 ] && echo "EMQX processes: $PS_LINE"
     if [ "$(echo -e "$PS_LINE" | wc -l)" -eq 1 ]; then
-        ## only one emqx node is running
-        ## strip 'emqx_data_dir ' and ' --' because the dir in between may contain spaces
-        DATA_DIR="$(echo -e "$PS_LINE" | grep -oE "\-emqx_data_dir.*" | sed -E 's#.+emqx_data_dir[[:blank:]]##g' | sed -E 's#[[:blank:]]--$##g' || true)"
-        if [ "$DATA_DIR" = '' ]; then
-            ## this should not happen unless -emqx_data_dir is not set
-            die "node_is_not_running!" 1
-        fi
-        # get ssl_dist_optfile option
-        SSL_DIST_OPTFILE="$(echo -e "$PS_LINE" | grep -oE '\-ssl_dist_optfile\s.+\s' | awk '{print $2}' || true)"
-        if [ -z "$SSL_DIST_OPTFILE" ]; then
-            EMQX_BOOT_CONFIGS="node.data_dir=${DATA_DIR}\ncluster.proto_dist=inet_tcp"
+        ## only one emqx node is running, get running args from 'ps -ef' output
+        tmp_nodename=$(echo -e "$PS_LINE" | grep -oE "\s\-s?name.*" | awk '{print $2}' || true)
```

Comment:
on macos `$2` gives `-name`, while `$3` works as expected - because of the whitespace I suppose
```
ps -ef | grep '[e]mqx' | grep -v -E '(remsh|nodetool)' | grep -oE "\-[r]oot /Users/id/work/emqx/_build/emqx/rel/emqx.*" |
grep -oE "\s\-s?name.*" | awk '{print $2}'
-name
```

```
ps -ef | grep '[e]mqx' | grep -v -E '(remsh|nodetool)' | grep -oE "\-[r]oot /Users/id/work/emqx/_build/emqx/rel/emqx.*" | grep -oE "\s\-s?name.*" | awk '{print $3}'
emqx@127.0.0.1
```

Reply:
hm, $2 works on my mac, version `13.0.1 (22A400)`

Reply:
for future readers: turns out my alias `grep='grep --color=always'` affected `awk`